### PR TITLE
feat: Support TypeScript moduleResolution nodenext option

### PR DESCRIPTION
**Pull request by CarlosLozanoHealthCaters**
TypeScript 4.7+ supports the new Node module resolution algorithm. More info: https://www.typescriptlang.org/docs/handbook/esm-node.html

However, Papr types doesn't work when using this new module resolution. The easiest way to make them work properly is by adding the extension '.d.ts' to all imports/exports in the generated type files. Similar to what we do for ESM files. I've modified the build script to fix this issue. I've checked the new generated files with different module resolutions and it was working. However, I encourage @avaly to recheck it.

I've also changed how the tsc command is invoked to guarantee that takes the version installed by the package.json and not the one installed in the global folder. In this way we keep consistency and we don't force to have typescript installed globally.

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -4,15 +4,17 @@ set -euxo pipefail
 
 # ESM
 
-tsc --project ./tsconfig-esm.json
+./node_modules/.bin/tsc --project ./tsconfig-esm.json
```
Comment (avaly) [inline build.sh:7]:
```suggestion
yarn tsc --project ./tsconfig-esm.json
```

File: build.sh [hunk 2]
```diff
@@ -4,15 +4,17 @@ set -euxo pipefail
...
 # This works in Linux and MacOS
 sed -i.backup -e "s/from '.\/\(.*\)';/from '.\/\1.js';/g" ./esm/*.js
+sed -i.backup -e "s/from '.\/\(.*\)';/from '.\/\1.d.ts';/g" ./esm/*.d.ts
 rm ./esm/*.js.backup
+rm ./esm/*.d.ts.backup
 
 # CommonJS
 
-tsc --project ./tsconfig-cjs.json
+./node_modules/.bin/tsc --project ./tsconfig-cjs.json
```
Comment (avaly) [inline build.sh:18]:
```suggestion
yarn tsc --project ./tsconfig-cjs.json
```

**Review by avaly — CHANGES_REQUESTED** [review]

**Comment by avaly** [comment]
Can you please add a new build test for this TS option?

File: build.sh (hunk 1 above)
Comment (CarlosLozanoHealthCaters) [reply build.sh]: I go back to the original version. I was executing ./build.sh instead of yarn build. It works correctly when executing yarn build. So I won't apply the change.

**Comment by CarlosLozanoHealthCaters** [comment]
@avaly Can you check if the test is done correctly?

I have also updated the MongoDB version in the tests to the version defined by Papr

File: tests/nodenext/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,12 @@
+{
+  "compilerOptions": {
+    "module": "NodeNext",
+    "target": "ES2022",
+    "outDir": "./dist",
+
```
Comment (avaly) [inline tests/nodenext/tsconfig.json:6]: nit: remove empty line and order the options alphabetically.

File: tests/nodenext/index.ts
Comment (avaly) [inline tests/nodenext/index.ts:1]: Please rename the folder as `ts-nodenext`, since this is related to the TS option.

File: .eslintignore [hunk 4]
```diff
@@ -2,4 +2,5 @@
 /cjs
 /docs
 /esm
-/tmp
\ No newline at end of file
+/tmp
+/tests/nodenext/index.ts
```
Comment (avaly) [inline .eslintignore:6]: Why is this needed here? We don't have entries for the other test files.

**Review by avaly — CHANGES_REQUESTED** [review]
LGTM overall.

File: tests/nodenext/index.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
+// This test does not use Jest because we want to test
+// that the output of the build process works under native Node.js ESM syntax
+// At this moment Jest does not support ESM syntax without transpiling
```
Comment (duncanbeevers) [inline tests/nodenext/index.ts:3]: Making the date and Jest version explicit in this comment may help future contributors.

File: tests/nodenext/tsconfig.json (hunk 3 above)
Comment (CarlosLozanoHealthCaters) [reply tests/nodenext/tsconfig.json]: done!

File: tests/nodenext/index.ts
Comment (CarlosLozanoHealthCaters) [reply tests/nodenext/index.ts:1]: done!

File: .eslintignore (hunk 4 above)
Comment (CarlosLozanoHealthCaters) [reply .eslintignore]: I was receiving a lint error in that file. I have changed the configuration of .eslintrc.json to include the tsconfig file in that folder. It seems to work.

File: tests/nodenext/index.ts (hunk 5 above)
Comment (CarlosLozanoHealthCaters) [reply tests/nodenext/index.ts]: @avaly I copied these comments from ESM test. Should we change both?

File: .eslintignore (hunk 4 above)
Comment (CarlosLozanoHealthCaters) [reply .eslintignore]: How can I fix that issue? My intellisense is getting the correct typings for papr.

File: tests/nodenext/index.ts (hunk 5 above)
Comment (CarlosLozanoHealthCaters) [reply tests/nodenext/index.ts]: done!

File: .eslintignore (hunk 4 above)
Comment (avaly) [reply .eslintignore]: Adding a top level `/* eslint-disable @typescript-eslint/no-unsafe-call */` in that file should be fine.

Comment (CarlosLozanoHealthCaters) [reply .eslintignore]: done! Thanks!

**Comment by avaly** [comment]
@CarlosLozanoHealthCaters Please rebase on the latest `main` commit and add the missing `expectType` calls from #461.

**Comment by CarlosLozanoHealthCaters** [comment]
Done! Rebased + Changes applied
